In [1]:
import pandas as pd

# Анализ данных об аренде велосипедов в Чикаго

In [10]:
bikes_q1 = pd.read_csv(bikes_q1, parse_dates=['start_time','end_time'] )

Для начала, возьмем данные только за Q1 <br>


### Структура данных
`trip_id` — id поездки <br>
`start_time` — Дата и время начала поездки<br>
`end_time` — Дата и время конца поездки<br>
`bikeid` — id велосипеда<br>
`tripduration` — продолжительность поездки в минутах<br>
`from_station_id` — id станции начала поездки<br>
`from_station_name` — название пункта отправления<br>
`to_station_id` — id пункта прибытия<br>
`to_station_name` — название пункта прибытия<br>
`usertype` — тип пользователя<br>
`gender` — пол (если subscriber)<br>
`birthyear` — год рождения (если subscriber)<br>

In [3]:
bikes_q1.head()

trip_id           start_time             end_time  bikeid tripduration  \
0  17617135  2018-01-22 20:04:31  2018-01-22 20:11:53    1131        442.0   
1  17897619  2018-03-16 19:47:59  2018-03-16 20:04:00    6146        961.0   
2  17881307  2018-03-14 18:49:20  2018-03-14 18:54:38    3847        318.0   
3  17881130  2018-03-14 18:33:48  2018-03-14 19:07:40    1483      2,032.0   
4  17686289  2018-02-05 17:39:14  2018-02-05 17:46:13    6391        419.0   

   from_station_id           from_station_name  to_station_id  \
0              471  Francisco Ave & Foster Ave            468   
1              296      Broadway & Belmont Ave            253   
2              260  Kedzie Ave & Milwaukee Ave            503   
3              199      Wabash Ave & Grand Ave            199   
4              596      Benson Ave & Church St            605   

               to_station_name    usertype  gender  birthyear  
0        Budlong Woods Library  Subscriber  Female     1949.0  
1  Winthrop Ave & Lawrence Ave  Subscriber    Male     1988.0  
2    Drake Ave & Fullerton Ave  Subscriber    Male     1987.0  
3       Wabash Ave & Grand Ave  Subscriber    Male     1990.0  
4      University Library (NU)  Subscriber    Male     1992.0

In [4]:
bikes_q1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 174215 entries, 0 to 174214
Data columns (total 12 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   trip_id            174215 non-null  int64  
 1   start_time         174215 non-null  object 
 2   end_time           174215 non-null  object 
 3   bikeid             174215 non-null  int64  
 4   tripduration       174215 non-null  object 
 5   from_station_id    174215 non-null  int64  
 6   from_station_name  174215 non-null  object 
 7   to_station_id      174215 non-null  int64  
 8   to_station_name    174215 non-null  object 
 9   usertype           174215 non-null  object 
 10  gender             164161 non-null  object 
 11  birthyear          164236 non-null  float64
dtypes: float64(1), int64(4), object(7)
memory usage: 15.9+ MB


In [6]:
bikes_q1.duplicated().sum()

0

In [8]:
bikes_q1.dtypes
# так как у столбцов с датой стоит тип object, приведем к формату DateTime при чтении csv файла

trip_id                int64
start_time            object
end_time              object
bikeid                 int64
tripduration          object
from_station_id        int64
from_station_name     object
to_station_id          int64
to_station_name       object
usertype              object
gender                object
birthyear            float64
dtype: object

#### Перед тем как сделать .resample(), немного подготовим данные. Поместим колонку start_time в качестве индексов и сохраните изменения в исходный датасет.

In [12]:
bikes_q1 = bikes_q1.set_index('start_time')

In [14]:
bikes_q1.head()

trip_id            end_time  bikeid tripduration  \
start_time                                                               
2018-01-22 20:04:31  17617135 2018-01-22 20:11:53    1131        442.0   
2018-03-16 19:47:59  17897619 2018-03-16 20:04:00    6146        961.0   
2018-03-14 18:49:20  17881307 2018-03-14 18:54:38    3847        318.0   
2018-03-14 18:33:48  17881130 2018-03-14 19:07:40    1483      2,032.0   
2018-02-05 17:39:14  17686289 2018-02-05 17:46:13    6391        419.0   

                     from_station_id           from_station_name  \
start_time                                                         
2018-01-22 20:04:31              471  Francisco Ave & Foster Ave   
2018-03-16 19:47:59              296      Broadway & Belmont Ave   
2018-03-14 18:49:20              260  Kedzie Ave & Milwaukee Ave   
2018-03-14 18:33:48              199      Wabash Ave & Grand Ave   
2018-02-05 17:39:14              596      Benson Ave & Church St   

                     to_station_id              to_station_name    usertype  \
start_time                                                                    
2018-01-22 20:04:31            468        Budlong Woods Library  Subscriber   
2018-03-16 19:47:59            253  Winthrop Ave & Lawrence Ave  Subscriber   
2018-03-14 18:49:20            503    Drake Ave & Fullerton Ave  Subscriber   
2018-03-14 18:33:48            199       Wabash Ave & Grand Ave  Subscriber   
2018-02-05 17:39:14            605      University Library (NU)  Subscriber   

                     gender  birthyear  
start_time                              
2018-01-22 20:04:31  Female     1949.0  
2018-03-16 19:47:59    Male     1988.0  
2018-03-14 18:49:20    Male     1987.0  
2018-03-14 18:33:48    Male     1990.0  
2018-02-05 17:39:14    Male     1992.0

#### В данных имеется как дата аренды, так и её точное время начала и окончания с точностью до секунд. Применим метод pd.resample() и агрегируем данные по дням. Посмотрим максимальное число аренд за день

In [23]:
bikes_q1.resample('1d').trip_id.count().sort_values(ascending=False)

start_time
2018-02-27    4196
2018-03-28    3640
2018-02-26    3422
2018-02-28    3412
2018-03-15    3245
              ... 
2018-02-10     364
2018-02-11     331
2018-02-09     300
2018-02-04     279
2018-01-01     191
Name: trip_id, Length: 90, dtype: int64

**Вывод:** максимально число аренд за день  было 27 февраля и составляло 4196